indico/modules/events/registration/clone.py

```diff
@@ -22,47 +22,85 @@
 from indico.modules.events.registration.models.forms import RegistrationForm
 from indico.modules.events.registration.models.form_fields import RegistrationFormFieldData
 from indico.modules.events.registration.models.items import RegistrationFormItem, RegistrationFormSection
+from indico.modules.events.registration.models.registrations import Registration, RegistrationData
 from indico.util.i18n import _
 
 from MaKaC.conference import EventCloner
 
 
 class RegistrationFormCloner(EventCloner):
+    def __init__(self, *args, **kwargs):
+        super(RegistrationFormCloner, self).__init__(*args, **kwargs)
+        self.field_data_dict = {}
+        self._clone_with_registrations = False
+
     def get_options(self):
         if not is_feature_enabled(self.event, 'registration'):
             return {}
         enabled = bool(self._find_registration_forms().count())
-        return {'form': (_('Registration forms'), enabled, False)}
+        return {
+            'form': (_('Registration forms'), enabled, False),
+            'registrations': (_('Registrations'), enabled, False)
+        }
+
+    def _find_registration_forms(self):
+        return RegistrationForm.find(~RegistrationForm.is_deleted,
+                                     RegistrationForm.event_id == int(self.event.id))
 
     def clone(self, new_event, options):
-        if 'form' not in options:
+        if not {'form', 'registrations'} & options:
             return
+        self._clone_with_registrations = 'registrations' in options
         attrs = get_simple_column_attrs(RegistrationForm) - {'start_dt', 'end_dt'}
         for old_form in self._find_registration_forms():
             new_form = RegistrationForm(event_id=new_event.id, **{attr: getattr(old_form, attr) for attr in attrs})
             self._clone_form_items(old_form, new_form)
+            if self._clone_with_registrations:
+                self._clone_registrations(old_form, new_form, new_event)
             db.session.add(new_form)
             db.session.flush()
 
-    def _find_registration_forms(self):
-        return RegistrationForm.find(~RegistrationForm.is_deleted,
-                                     RegistrationForm.event_id == int(self.event.id))
-
     def _clone_form_items(self, old_form, new_form):
         old_sections = RegistrationFormSection.find(RegistrationFormSection.registration_form_id == old_form.id,
                                                     ~RegistrationFormSection.is_deleted)
```

Comment:
When cloning registrations we need to include deleted sections/fields, too (or check whether any of the registrations have references to deleted fields or fields in deleted sections, but I think we don't really need that - cloning the deleted ones in any case when cloning registrations should be fine)